[SVG2] Sync `getTotalLength()` with web specification to throw exception when non-renderable and path is empty

Ahmad-S792 opened the pull request:
#### 48ed528d4e27ad7f8424955add1d503ed28f0ee6
<pre>
[SVG2] Sync `getTotalLength()` with web specification to throw exception when non-renderable and path is empty

<a href="https://bugs.webkit.org/show_bug.cgi?id=280350">https://bugs.webkit.org/show_bug.cgi?id=280350</a>
<a href="https://rdar.apple.com/136719548">rdar://136719548</a>

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Web Specification [1]:

[1] <a href="https://svgwg.org/svg2-draft/types.html#__svg__SVGGeometryElement__getTotalLength">https://svgwg.org/svg2-draft/types.html#__svg__SVGGeometryElement__getTotalLength</a>

&quot;If current element is a non-rendered element, and the UA is not able to
compute the total length of the path, then throw an InvalidStateError.&quot;

In following patch, we update to throw exception when there is no `renderer`
and also when path is empty similar to `getPointAtLength`.

It merges tests from following Chromium commit [2]:

[2] <a href="https://source.chromium.org/chromium/chromium/src/+/5ad2c261332caaad06a0827addccb575215e87b7">https://source.chromium.org/chromium/chromium/src/+/5ad2c261332caaad06a0827addccb575215e87b7</a>

* Source/WebCore/svg/SVGGeometryElement.cpp:
(WebCore::SVGGeometryElement::getTotalLength const):
(WebCore::SVGGeometryElement::getPointAtLength const):
* Source/WebCore/svg/SVGGeometryElement.h:
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::getTotalLength const):
(WebCore::SVGPathElement::getPointAtLength const):
* Source/WebCore/svg/SVGPathElement.h:
* LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-exception.html:
* LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-exception-expected.txt:
* LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-detached.html:
* LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-detached-expected.txt:

Canonical link: <a href="https://commits.webkit.org/284311@main">https://commits.webkit.org/284311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ba82449b70cfab093b6d12b02603c241ee5141e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69013 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21685 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73093 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/20167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71130 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56214 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20019 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54956 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/20167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72079 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44200 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59580 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35430 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40865 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17009 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18545 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62814 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17355 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74802 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12995 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16602 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62598 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13033 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59663 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62495 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15317 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10482 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4088 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44217 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45290 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46486 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45032 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-exception-expected.txt:
```diff
@@ -0,0 +1,4 @@
+
+PASS SVGGeometryElement-getTotalLength-exception
+PASS SVGGeometryElement-getTotalLength-exception 1
```
Where does this `1` come from?

Ahmad-S792 replied:
Updated, it was coming because it was taking test name and both test had no explicit name, so '1' was getting appended in front of it.

shallawa commented on Source/WebCore/svg/SVGGeometryElement.cpp:
```diff
@@ -46,13 +46,13 @@ SVGGeometryElement::SVGGeometryElement(const QualifiedName& tagName, Document& d
     auto* renderer = this->renderer();
     if (!renderer)
-        return 0;
+        return Exception { ExceptionCode::InvalidStateError };
```
Should not an exception message be included here?
```
return Exception { ExceptionCode::InvalidStateError, "The SVG element can not be rendered."_s  };
```

Ahmad-S792 replied:
Yes - can amend it. :-)

shallawa commented on Source/WebCore/svg/SVGGeometryElement.cpp:
```diff
@@ -74,7 +74,10 @@ ExceptionOr<Ref<SVGPoint>> SVGGeometryElement::getPointAtLength(float distance)
+    auto totalLength =  getTotalLength();
+    if (totalLength.hasException())
+        return totalLength.releaseException();
+    distance = clampTo<float>(distance, 0, totalLength.releaseReturnValue());
```
`totalLength` is just float so it does not need the move semantic. So I think we can use `totalLength.returnValue()`.

Ahmad-S792 replied:
Done and also updated it for `SVGPathElement`.

shallawa commented on LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-detached.html:
```diff
@@ -0,0 +1,43 @@
+
+  assert_equals(getTotalLength('M0,20 L400,20 L640,20'), 640);
+  assert_equals(getTotalLength('M0,20 L400,20 L640,20 z'), 1280);
+  assert_equals(getTotalLength('M0,20 L400,20 z M 320,20 L640,20'), 1120);
```
Can we add a test for the empty path case:
```
assert_throws_dom("InvalidStateError", function() { getTotalLength('') });
```

Ahmad-S792 replied:
Added.

shallawa commented on Source/WebCore/svg/SVGPathElement.cpp:
```diff
@@ -194,7 +197,10 @@ ExceptionOr<Ref<SVGPoint>> SVGPathElement::getPointAtLength(float distance) cons
         return Exception { ExceptionCode::InvalidStateError, "The element's path is empty."_s };
 
     // Spec: Clamp distance to [0, length].
```
Please move this comment to where the clamping happens:
```
    auto totalLength =  getTotalLength();
    if (totalLength.hasException())
        return totalLength.releaseException();

    // Spec: Clamp distance to [0, length].
    distance = clampTo<float>(distance, 0, totalLength.releaseReturnValue());
```

Ahmad-S792 replied:
Will update. :-)

Ahmad-S792 replied:
Done for both.

Ahmad-S792 committed:
[SVG2] Sync `getTotalLength()` with web specification to throw exception when non-renderable and path is empty

https://bugs.webkit.org/show_bug.cgi?id=280350
rdar://136719548

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Web Specification [1]:

[1] https://svgwg.org/svg2-draft/types.html#__svg__SVGGeometryElement__getTotalLength

"If current element is a non-rendered element, and the UA is not able to
compute the total length of the path, then throw an InvalidStateError."

In following patch, we update to throw exception when there is no `renderer`
and also when path is empty similar to `getPointAtLength`.

It merges tests from following Chromium commit [2]:

[2] https://source.chromium.org/chromium/chromium/src/+/5ad2c261332caaad06a0827addccb575215e87b7

* Source/WebCore/svg/SVGGeometryElement.cpp:
(WebCore::SVGGeometryElement::getTotalLength const):
(WebCore::SVGGeometryElement::getPointAtLength const):
* Source/WebCore/svg/SVGGeometryElement.h:
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::getTotalLength const):
(WebCore::SVGPathElement::getPointAtLength const):
* Source/WebCore/svg/SVGPathElement.h:
* LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-exception.html:
* LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-exception-expected.txt:
* LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-detached.html:
* LayoutTests/svg/dom/SVGGeometryElement-getTotalLength-detached-expected.txt:

Canonical link: https://commits.webkit.org/284311@main
